packages/walletconnect/src/index.ts

```diff
@@ -1,121 +1,121 @@
-      return accumulator
-    }, {})
-    this.defaultChainId = defaultChainId ?? Number(Object.keys(this.rpc)[0])
+    this.chains = chains
+    this.defaultChainId = defaultChainId
+    this.rpcMap = rpcMap || rpc
+
     this.timeout = timeout
   }
 
-  private disconnectListener = (error?: ProviderRpcError): void => {
+  private disconnectListener = (error: ProviderRpcError) => {
     this.actions.resetState()
     if (error) this.onError?.(error)
   }
 
-  private chainChangedListener = (chainId: number | string): void => {
-    this.actions.update({ chainId: parseChainId(chainId) })
+  private chainChangedListener = (chainId: string): void => {
+    this.actions.update({ chainId: Number.parseInt(chainId, 16) })
   }
 
   private accountsChangedListener = (accounts: string[]): void => {
     this.actions.update({ accounts })
   }
 
-  private URIListener = (_: Error | null, payload: { params: string[] }): void => {
-    this.events.emit(URI_AVAILABLE, payload.params[0])
+  private URIListener = (uri: string): void => {
+    this.events.emit(URI_AVAILABLE, uri)
   }
 
-  private async isomorphicInitialize(chainId = this.defaultChainId): Promise<void> {
-    if (this.eagerConnection) return
-
-    // because we can only use 1 url per chainId, we need to decide between multiple, where necessary
-    const rpc = Promise.all(
-      Object.keys(this.rpc).map(
-        async (chainId): Promise<[number, string]> => [
-          Number(chainId),
-          await getBestUrl(this.rpc[Number(chainId)], this.timeout),
-        ]
-      )
-    ).then((results) =>
-      results.reduce<{ [chainId: number]: string }>((accumulator, [chainId, url]) => {
-        accumulator[chainId] = url
-        return accumulator
-      }, {})
-    )
-
-    return (this.eagerConnection = import('@walletconnect/ethereum-provider').then(async (m) => {
-      this.provider = new m.default({
+  private isomorphicInitialize(
+    desiredChainId: number | undefined = this.defaultChainId
+  ): Promise<WalletConnectProvider> {
+    if (this.eagerConnection) return this.eagerConnection
+
+    const rpcMap = this.rpcMap ? getBestUrlMap(this.rpcMap, this.timeout) : undefined
+    const chains = desiredChainId ? getChainsWithDefault(this.chains, desiredChainId) : this.chains
+
+    return (this.eagerConnection = import('@walletconnect/ethereum-provider').then(async (ethProviderModule) => {
+      const provider = (this.provider = await ethProviderModule.default.init({
         ...this.options,
-        chainId,
-        rpc: await rpc,
-      }) as unknown as MockWalletConnectProvider
-
-      this.provider.on('disconnect', this.disconnectListener)
-      this.provider.on('chainChanged', this.chainChangedListener)
-      this.provider.on('accountsChanged', this.accountsChangedListener)
-      this.provider.connector.on('display_uri', this.URIListener)
+        chains,
+        rpcMap: await rpcMap,
+      }))
+
+      provider.on('disconnect', this.disconnectListener)
```

Comment:
nit: I thought you were going to chain all of these things.

Reply:
Yes, done